src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Repository/FilterRepository.cs

```diff
@@ -0,0 +1,178 @@
+
+public class FilterRepository(
+    PublicDataDbContext publicDataDbContext,
+    IDataSetVersionPathResolver dataSetVersionPathResolver) : IFilterRepository
+{
+    public async Task CreateFilterMetas(
+        IDuckDbConnection duckDbConnection,
+        DataSetVersion dataSetVersion,
+        IReadOnlySet<string> allowedColumns,
+        CancellationToken cancellationToken)
+    {
+        var metas = (await duckDbConnection.SqlBuilder(
+                    $"""
+                     SELECT *
+                     FROM '{dataSetVersionPathResolver.CsvMetadataPath(dataSetVersion):raw}'
+                     WHERE "col_type" = {MetaFileRow.ColumnType.Filter.ToString()}
+                     AND "col_name" IN ({allowedColumns})
+                     """)
+                .QueryAsync<MetaFileRow>(cancellationToken: cancellationToken)
+            )
+            .OrderBy(row => row.Label)
+            .Select(
+                row => new FilterMeta
+                {
+                    PublicId = row.ColName,
+                    DataSetVersionId = dataSetVersion.Id,
+                    Label = row.Label,
+                    Hint = row.FilterHint ?? string.Empty,
+                }
+            )
+            .ToList();
+
+        publicDataDbContext.FilterMetas.AddRange(metas);
+        await publicDataDbContext.SaveChangesAsync(cancellationToken);
+
+        foreach (var meta in metas)
+        {
+            var options = (await duckDbConnection.SqlBuilder(
+                    $"""
+                     SELECT DISTINCT "{meta.PublicId:raw}"
+                     FROM read_csv('{dataSetVersionPathResolver.CsvDataPath(dataSetVersion):raw}', ALL_VARCHAR = true) AS data
+                     WHERE "{meta.PublicId:raw}" != ''
+                     ORDER BY "{meta.PublicId:raw}"
+                     """
+                ).QueryAsync<string>(cancellationToken: cancellationToken))
+                .Select(
+                    label => new FilterOptionMeta
+                    {
+                        Label = label, IsAggregate = label == "Total" ? true : null
+                    }
+                )
+                .ToList();
+
+            var optionTable = publicDataDbContext.GetTable<FilterOptionMeta>();
+
+            // Merge to only inserting new filter options
+            // that don't already exist in the table.
+            await optionTable
+                .Merge()
+                .Using(options)
+                .On(
+                    o => new
+                    {
+                        o.Label, o.IsAggregate
+                    },
+                    o => new
+                    {
+                        o.Label, o.IsAggregate
+                    }
+                )
+                .InsertWhenNotMatched()
+                .MergeAsync(cancellationToken);
+
+            var startIndex = await publicDataDbContext.FilterOptionMetaLinks.CountAsync(token: cancellationToken);
```

Comment:
I've added coordination now by acquiring a lock on a durable entity where the activity name is the entity id.

Prior to calling the `ImportMetadata` activity, the orchestration client will wait until it can obtain a lock on an entity with that name.

No other orchestrations can execute an activity concurrently provided that they also acquire a lock through the same process, i.e. the new method `TaskOrchestrationContextExtensions.CallActivityExclusively`.